plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java

```diff
@@ -294,6 +314,73 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         return Optional.of(new ConstraintApplicationResult<>(handle, remainingFilter, false));
     }
 
+    @Override
+    public Optional<AggregationApplicationResult<ConnectorTableHandle>> applyAggregation(
+            ConnectorSession session,
+            ConnectorTableHandle handle,
+            List<AggregateFunction> aggregates,
+            Map<String, ColumnHandle> assignments,
+            List<List<ColumnHandle>> groupingSets)
+    {
+        if (!isAggregationPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        // Global aggregation is represented by [[]]
+        verify(!groupingSets.isEmpty(), "No grouping sets provided");
+
+        // Pinot currently only supports simple GROUP BY clauses with a single grouping set
+        if (groupingSets.size() != 1) {
+            return Optional.empty();
+        }
+
+        PinotTableHandle tableHandle = (PinotTableHandle) handle;
+        // If aggregate are present than no further aggregations
+        // can be pushed down: there are currently no subqueries in pinot
+        if (tableHandle.getQuery().isPresent() &&
+                !tableHandle.getQuery().get().getAggregateColumns().isEmpty()) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<ConnectorExpression> projections = ImmutableList.builder();
+        ImmutableList.Builder<Assignment> resultAssignments = ImmutableList.builder();
+        ImmutableList.Builder<PinotColumnHandle> aggregationExpressions = ImmutableList.builder();
+
+        for (AggregateFunction aggregate : aggregates) {
+            Optional<PinotColumnHandle> rewriteResult = aggregateFunctionRewriter.rewrite(session, aggregate, assignments, tableHandle);
+            if (rewriteResult.isEmpty()) {
+                return Optional.empty();
+            }
+            PinotColumnHandle pinotColumnHandle = rewriteResult.get();
+            aggregationExpressions.add(pinotColumnHandle);
+            projections.add(new Variable(pinotColumnHandle.getColumnName(), pinotColumnHandle.getDataType()));
+            resultAssignments.add(new Assignment(pinotColumnHandle.getColumnName(), pinotColumnHandle, pinotColumnHandle.getDataType()));
+        }
+        List<String> groupingColumns = getOnlyElement(groupingSets).stream()
+                .map(PinotColumnHandle.class::cast)
+                .map(PinotColumnHandle::getColumnName)
+                .collect(toImmutableList());
+        OptionalLong limitForDynamicTable = OptionalLong.empty();
+        // Ensure that pinot default limit of 10 rows is not used
+        if (tableHandle.getLimit().isEmpty() && !groupingColumns.isEmpty()) {
+            limitForDynamicTable = OptionalLong.of(maxRowsPerBrokerQuery + 1);
```

Comment:
Why the +1? Deserves a comment in my opinion since it's not obvious (to me at least).

Reply:
Sure will add one. Just for context: this is similar to the limit for segment queries which was done like the elastic search connector, @martint pointed me to that code: if the limit is + 1 then the connector will know when too many rows are returned and throw an exception.